common/chains/client/types.go

```diff
@@ -0,0 +1,126 @@
+package client
+
+import (
+	"context"
+	"math/big"
+
+	feetypes "github.com/smartcontractkit/chainlink/v2/common/fee/types"
+	"github.com/smartcontractkit/chainlink/v2/common/types"
+	"github.com/smartcontractkit/chainlink/v2/core/assets"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+// RPC includes all the necessary methods for a multi-node client to interact directly with any RPC endpoint.
+type RPC[
+	CHAIN_ID types.ID,
+	SEQ types.Sequence,
+	ADDR types.Hashable,
+	BLOCK_HASH types.Hashable,
+	TX any,
+	TX_HASH types.Hashable,
+	EVENT any,
+	EVENT_OPS any,
+	TX_RECEIPT types.Receipt[TX_HASH, BLOCK_HASH],
+	FEE feetypes.Fee,
+	HEAD types.Head[BLOCK_HASH],
+
+] interface {
+	NodeClient[
+		CHAIN_ID,
+		HEAD,
+	]
+	ClientAPI[
+		CHAIN_ID,
+		SEQ,
+		ADDR,
+		BLOCK_HASH,
+		TX,
+		TX_HASH,
+		EVENT,
+		EVENT_OPS,
+		TX_RECEIPT,
+		FEE,
+		HEAD,
+	]
+}
+
+// Head is the interface required by the NodeClient
+type Head interface {
+	BlockNumber() int64
+	BlockDifficulty() *utils.Big
+}
+
+// NodeClient includes all the necessary RPC methods required by a node.
+type NodeClient[
+	CHAIN_ID types.ID,
+	HEAD Head,
+] interface {
+	Close()
+	ChainID(context.Context) (CHAIN_ID, error)
+	Dial(callerCtx context.Context) error
+	DialHTTP() error
+	DisconnectAll()
+	Subscribe(ctx context.Context, channel chan<- HEAD, args ...interface{}) (types.Subscription, error)
+	ClientVersion(context.Context) (string, error)
```

Comment:
I see some overlap with `ClientAPI`. Have these been fully simplified? Would it be helpful to have another type embedded in both for the comon ones?
Otherwise, maybe just align the grouping:
```suggestion
	// Connection
	ClientVersion(context.Context) (string, error)
	Close()
	Dial(ctx context.Context) error
	DialHTTP() error
	DisconnectAll()

	// Misc
	ChainID(context.Context) (CHAIN_ID, error)
	Subscribe(ctx context.Context, channel chan<- HEAD, args ...interface{}) (types.Subscription, error)
```

Reply:
TBH it felt better to keep some duplicates for better clarity. I wanted the reader to be able to understand what each interface needs at a quick glance. Dropping duplicates or adding an intermediate interface felt like it would hurt readability 🤷